microprofile/grpc/core/src/main/java/io/helidon/microprofile/grpc/core/AbstractServiceBuilder.java

```diff
@@ -0,0 +1,248 @@
+     *
+     * @return  {@code true} if this modeller contains an annotated service
+     */
+    public boolean isAnnotatedService() {
+        return annotatedServiceClass.isAnnotationPresent(Grpc.class);
+    }
+
+    /**
+     * Obtain the service class.
+     * @return  the service class
+     */
+    protected Class<?> serviceClass() {
+        return serviceClass;
+    }
+
+    /**
+     * Obtain the actual annotated class.
+     * @return  the actual annotated class
+     */
+    protected Class<?> annotatedServiceClass() {
+        return annotatedServiceClass;
+    }
+
+    /**
+     * Obtain the {@link MarshallerSupplier} to use.
+     * <p>
+     * The {@link MarshallerSupplier} will be determined by the {@link GrpcMarshaller}
+     * annotation if it is present otherwise the default supplier will be returned.
+     *
+     * @return  the {@link MarshallerSupplier} to use
+     */
+    protected MarshallerSupplier getMarshallerSupplier() {
+        GrpcMarshaller annotation = annotatedServiceClass.getAnnotation(GrpcMarshaller.class);
+        return annotation == null ? MarshallerSupplier.create() : ModelHelper.getMarshallerSupplier(annotation);
+    }
+
+    /**
+     * Create the service instance supplier.
+     *
+     * @param cls  the service class
+     * @return the service instance supplier
+     */
+    protected static Supplier<?> createInstanceSupplier(Class<?> cls) {
+        if (cls.isAnnotationPresent(Singleton.class)) {
+            return Instance.singleton(cls);
+        } else {
+            return Instance.create(cls);
+        }
+    }
+
+    /**
+     * Verify that there are no non-public annotated methods.
+     */
+    protected void checkForNonPublicMethodIssues() {
+        AnnotatedMethodList allDeclaredMethods = AnnotatedMethodList.create(getAllDeclaredMethods(serviceClass));
+
+        // log warnings for all non-public annotated methods
+        allDeclaredMethods.withMetaAnnotation(GrpcMethod.class).isNotPublic()
+                .forEach(method -> LOGGER.log(Level.WARNING, () -> String.format("The gRPC method, %s, MUST be "
+                                              + "public scoped otherwise the method is ignored", method)));
+    }
+
+    /**
+     * Obtain the list of method handler suppliers.
+     *
+     * @return the list of method handler suppliers
+     */
+    protected List<MethodHandlerSupplier> handlerSuppliers() {
+        return handlerSuppliers;
+    }
+
+    /**
+     * Obtain the service instance supplier.
+     *
+     * @return the service instance supplier
+     */
+    protected Supplier<?> instanceSupplier() {
+        return instance;
+    }
+
+    /**
+     * Obtain a list of all of the methods declared on the service class.
+     *
+     * @param clazz  the service class
+     * @return a list of all of the methods declared on the service class
+     */
+    protected List<Method> getAllDeclaredMethods(Class<?> clazz) {
+        List<Method> result = new LinkedList<>();
```

Comment:
So much reflection at runtime. Could this be handle by a CDI extension at least?